### DISCUSSION TITLE
How to select record that belongs to currently auth'd user?

### DISCUSSION THREAD

**Discussion by renefournier** [supabase/supabase#4634] [Questions]
(Warning, I'm a newbie. I've read the docs, but still unsure how to structure some code.)

In a nutshell, in my SvelteKit app, user logs in (auth), gets and sets a session cookie (email), stores the accessToken in a store, is redirected to their Profile page.... And when it mounts, the app should fetch the profile record in the DB that belongs to them—not all of them. Here's my code that works, but you'll notice the `id` is hard-coded:

```
	let profile = {};

	onMount(async () => {
		console.log('🐎');
		supabase.auth.setAuth($accessToken);

		console.dir(session);
		console.dir($accessToken);

		let { data: profiles, error } = await supabase
			.from('profiles')
			.select(
				'id, address, clinic, location, owner, subscription_expires, tel, url_id, website, welcome'
			)
			.eq('id', '7419a496-7b4f-49e1-8284-ddae247b258e');

		if (error) {
			console.warn(error);
		}
		if (!profiles || profiles.length === 0) {
			waitingForNetwork = false;
			alert('Hmm, no profile data');
			goto('/api/logout');
		} else {
			waitingForNetwork = false;
			profile = profiles[0];
		}
	});
```

Any suggestions?

**Comment by GaryAustin1** [comment]
Edit: For future users finding this, my comment here was wrong.  setAuth returns a session object with null for user.  It does not make a server call, just sets the jwt to allow access.  If you need the user info on a server then getUser(token) is the call needed.

You should be able to get the user object with const { user, error } = supabase.auth.setAuth($access_token).
BUT if you want users to only access their own profile, you really need to use RLS to limit the table access with auth.uid().

**Reply by renefournier** [reply]
I probably should explain it better. I do have RLS policies that only users owning the profile `(uid() = id)` can update it. And all users can view all rows (`true`). BUT, when the user signs in and visits his `/profile` page, I only want returned his profile, not an array of all profiles. 

I'm wondering, when `auth` happens, should I perhaps store the user's UUID (not just email) in the session in order to use in the query, e.g.:

```
let { data: profiles, error } = await supabase
    .from('profiles')
    .select('id, address, clinic, location, owner, subscription_expires, tel, url_id, website, welcome')
    .eq('id', sessionId);
```

**Reply by GaryAustin1** [reply]
The first sentence in my answer is how you can get the UUID when you call setAuth.  Then you should have user and user.id

**Reply by renefournier** [reply]
> You should be able to get the user object with const { user, error } = supabase.auth.setAuth($access_token). BUT if you want users to only access their own profile, you really need to use RLS to limit the table access with auth.uid().

When I call `setAuth` nothing is returned.

```
  onMount(async () => {
    console.log('🐎');
    const { user, error } = supabase.auth.setAuth($accessToken);

    console.dir(user);
    console.dir(error);
    console.dir($accessToken);
...
```
In the `console` I see:
```
🐎
profile.svelte:87 null
profile.svelte:88 undefined
profile.svelte:89 eyJhbGciOiJIUzI...
```
Strange, no? I must be holding it wrong...

**Reply by GaryAustin1** [reply]
I’m somewhat confused on setAuth().  In looking at its code it just sets the token into the session and then apparently sets user to null.   I guess it is resetting the session and you then have to make a call like 
 const user = await supabase.auth.user();
 to get the new user info using the new token.
Edit: but if you already have the ID from previous operations putting it in a store should work also.

**Reply by renefournier** [reply]
> I’m somewhat confused on setAuth(). In looking at its code it just sets the token into the session and then apparently sets user to null. I guess it is resetting the session and you then have to make a call like const user = await supabase.auth.user(); to get the new user info using the new token. Edit: but if you already have the ID from previous operations putting it in a store should work also.

Hmm, user (`u`) is still `null`.

```	onMount(async () => {
		console.log('🐎');
		const { user, error } = supabase.auth.setAuth($accessToken);

		console.dir(user);
		console.dir(error);
		console.dir($accessToken);

		const u = await supabase.auth.user();
		console.dir(u);
```
<img width="494" alt="Screen Shot 2021-12-23 at 4 13 19 PM" src="https://user-images.githubusercontent.com/4491052/147298639-8669d5c5-2595-4964-ab96-3a909cb2ffcc.png">

**Reply by GaryAustin1** [reply]
I’m going to guess your supabase session is not set up when you do this mount or is lost from server to client.  I use sveltekit but with static pages so don’t have any server operations.  I assume since you are passing a token around you do.

Edit: At this point I hope some more familiar with use of tokens comes in with some help and I’m happy to delete this answer if it gives you a better shot at getting noticed.

**Reply by renefournier** [reply]
The session appears to be set up correctly from server to client, or at least server, because the "not auth" redirect does not run. I'm using SvelteKit and deploying to Netlify, and this problem exists in both `dev` and `prod`. After logging-in, I do have a session cookie with my email, and do `setAuth` with the accessToken that pass from server to client via HTTP header seems to do its thing. But... I guess I don't understand how Supabase's client juggles the session cookie and JWT. I love SvelteKit, but the SSR stuff makes this kinda tricky. Thanks for helping. Think I might punt my dumb question to @kiwicopple or one of the Supabase dudes.

**Reply by GaryAustin1** [reply]
Silentworks one of the supabase discord mods is pretty good with sveltekit stuff.

**Comment by silentworks** [comment]
Your code here looks like it's following the Svelte example rather than how you would handle this in SvelteKit. I have an example app I created which is using SvelteKit and Supabase along with getting a user's profile at the same time, take a look to see if it helps you. You need to do both client and server side to get the authentication working as it should, so you have to make a request from the client to the server to start the auth process for the server. https://github.com/silentworks/waiting-list/blob/main/src/routes/__layout.svelte#L19

You need to have `supabase.auth.onAuthStateChange` running in your code.

https://github.com/silentworks/waiting-list

**Comment by j314h** [comment]
hello, i'm new to supabase, but i'm familiar with sveltekit and i understand @silentworks answer and i think he's right.  but I have a question because I am having difficulties with Supabase's Real Time with sveltekit kit, because as you say you have to go through the server and there things are complicated.  did you find a solution?  or test realtime with RLS and sveltekit?  i need your help please
